mne/preprocessing/xdawn.py

```diff
@@ -1,266 +1,332 @@
+        if float, shrinkage covariance is used (0 <= shrinkage <= 1).
+        if str, optimal shrinkage using Ledoit-Wolf Shrinkage ('ledoit_wolf')
+        or Oracle Approximating Shrinkage ('oas').
+    signal_cov : None | Covariance | array, shape (n_channels, n_channels)
+        The signal covariance used for whitening of the data.
+        if None, the covariance is estimated from the epochs signal.
+    events : array, shape (n_epochs, 3)
+        The epochs events, used to correct for epochs overlap.
+    tmin : float
+        Epochs starting time. Only used if events is passed to correct for
+        epochs overlap.
+    sfreq : float
+        Sampling frequency.  Only used if events is passed to correct for
+        epochs overlap.
 
-def _construct_signal_from_epochs(events, sfreq, tmin, tmax, epochs_data):
-    """Reconstruct pseudo continuous signal from epochs."""
-    start = (np.min(events[:, 0]) + int(tmin * sfreq))
-    stop = (np.max(events[:, 0]) + int(tmax * sfreq) + 1)
 
-    n_samples = stop - start
-    n_epochs, n_channels, n_times = epochs_data.shape
-    events_pos = events[:, 0] - events[0, 0]
+    Attributes
+    ----------
+    filters_ : array, shape (n_channels, n_channels)
+        The Xdawn components used to decompose the data for each event type.
+    patterns_ : array, shape (n_channels, n_channels)
+        The Xdawn patterns used to restore the signals for each event type.
 
-    data = np.zeros((n_channels, n_samples))
-    for idx in range(n_epochs):
-        onset = events_pos[idx]
-        offset = onset + n_times
-        data[:, onset:offset] = epochs_data[idx]
+    References
+    ----------
+    [1] Rivet, B., Souloumiac, A., Attina, V., & Gibert, G. (2009). xDAWN
+    algorithm to enhance evoked potentials: application to brain-computer
+    interface. Biomedical Engineering, IEEE Transactions on, 56(8), 2035-2043.
+    [2] Rivet, B., Cecotti, H., Souloumiac, A., Maby, E., & Mattout, J. (2011,
+    August). Theoretical analysis of xDAWN algorithm: application to an
+    efficient sensor selection in a P300 BCI. In Signal Processing Conference,
+    2011 19th European (pp. 1382-1386). IEEE.
 
-    return data
 
+    See Also
+    --------
+    CSP
+    """
+    n_epochs, n_channels, n_times = epochs.shape
+
+    classes = np.unique(y)
+
+    # Retrieve or compute whitening covariance
+    if signal_cov is None:
+        signal_cov = _regularized_covariance(np.hstack(epochs), reg)
+    elif isinstance(signal_cov, Covariance):
+        signal_cov = signal_cov.data
+    if not isinstance(signal_cov, np.ndarray) or (
+            not np.array_equal(signal_cov.shape, np.tile(epochs.shape[1], 2))):
+        raise ValueError('signal_cov must be None, a covariance instance '
+                         'or a array of shape (n_chans, n_chans)')
+
+    # Get prototype events
+    if events is not None:
+        evokeds, toeplitzs = _least_square_evoked(epochs, events, tmin, sfreq)
```

Comment:
With the current code, it would be something like

```
from mne.stats.regression import _prepare_rerp_preds, _make_evokeds
from mne.preprocessing.xdawn import _construct_signal_from_epochs
from sklearn.linear_model import Ridge
raw_, events_ = _construct_signal_from_epochs(events, epochs.info["sfreq"], tmin, tmax, epochs.get_data())  # function would have to be modified to return events)
toeplitzs, conds, cond_length, tmin_s, tmax_s = _prepare_rerp_preds(
    raw_.n_times, raw_.info["sfreq"], events_, event_id, tmin, tmax)
coefs = Ridge(alpha=0).fit(toeplitzs, raw_._data).coefs_
evokeds = _make_evokeds(coefs, conds, cond_length, tmin_s, tmax_s, raw_.info)
```

Hm. Doesn't look that hard actually.
I think I'll try finally doing this (1 year after I opened the issue) once this one is merged.

To build the evokeds, you need some information about the delays/estimated response functions per condition to extract the data from coefs. So if you use a different mechanism to build X/toeplitzs, that will have to be taken care of.